File: bridge/SfxrNode.cpp
#include "./dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
#include "struct.h"
using namespace lab;

DART_EXPORT int createSfxrNode(AudioContext* context) {
    auto node = std::make_shared<SfxrNode>(*context);
    return keepNode(node);
}

// todo AudioSetting: preset waveType  

DART_EXPORT int SfxrNode_attackTime(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 1, node->attackTime()) : -1;
}
DART_EXPORT int SfxrNode_sustainTime(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 2, node->sustainTime()) : -1;
}
DART_EXPORT int SfxrNode_sustainPunch(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 3, node->sustainPunch()) : -1;
}
DART_EXPORT int SfxrNode_decayTime(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 4, node->decayTime()) : -1;
}

DART_EXPORT int SfxrNode_startFrequency(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 5, node->startFrequency()) : -1;
}
DART_EXPORT int SfxrNode_minFrequency(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 6, node->minFrequency()) : -1;
}
DART_EXPORT int SfxrNode_slide(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 7, node->slide()) : -1;
}
DART_EXPORT int SfxrNode_deltaSlide(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 8, node->deltaSlide()) : -1;
}
DART_EXPORT int SfxrNode_vibratoDepth(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 9, node->vibratoDepth()) : -1;
}
DART_EXPORT int SfxrNode_vibratoSpeed(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 10, node->vibratoSpeed()) : -1;
}

DART_EXPORT int SfxrNode_changeAmount(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 11, node->changeAmount()) : -1;
}
DART_EXPORT int SfxrNode_changeSpeed(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 12, node->changeSpeed()) : -1;
}

DART_EXPORT int SfxrNode_squareDuty(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 13, node->squareDuty()) : -1;
}
DART_EXPORT int SfxrNode_dutySweep(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 14, node->dutySweep()) : -1;
}

DART_EXPORT int SfxrNode_repeatSpeed(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 15, node->repeatSpeed()) : -1;
}
DART_EXPORT int SfxrNode_phaserOffset(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 16, node->phaserOffset()) : -1;
}
DART_EXPORT int SfxrNode_phaserSweep(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 17, node->phaserSweep()) : -1;
}


DART_EXPORT int SfxrNode_lpFilterCutoff(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 18, node->lpFilterCutoff()) : -1;
}
DART_EXPORT int SfxrNode_lpFilterCutoffSweep(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 19, node->lpFilterCutoffSweep()) : -1;
}
DART_EXPORT int SfxrNode_lpFiterResonance(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 20, node->lpFiterResonance()) : -1;
}
DART_EXPORT int SfxrNode_hpFilterCutoff(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 21, node->hpFilterCutoff()) : -1;
}
DART_EXPORT int SfxrNode_hpFilterCutoffSweep(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 22, node->hpFilterCutoffSweep()) : -1;
}


DART_EXPORT void SfxrNode_setStartFrequencyInHz(int nodeId, float value) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node) node->setStartFrequencyInHz(value);
}
DART_EXPORT void SfxrNode_setVibratoSpeedInHz(int nodeId, float value) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node) node->setVibratoSpeedInHz(value);
}

DART_EXPORT float SfxrNode_envelopeTimeInSeconds(int nodeId, float sfxrEnvTime) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? node->envelopeTimeInSeconds(sfxrEnvTime) : 0.0;
}
DART_EXPORT float SfxrNode_envelopeTimeInSfxrUnits(int nodeId, float t) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? node->envelopeTimeInSfxrUnits(t) : 0.0;
}
DART_EXPORT float SfxrNode_frequencyInSfxrUnits(int nodeId, float hz) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? node->frequencyInSfxrUnits(hz) : 0.0;
}
DART_EXPORT float SfxrNode_frequencyInHz(int nodeId, float sfxr) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? node->frequencyInHz(sfxr) : 0.0;
}
DART_EXPORT float SfxrNode_vibratoInSfxrUnits(int nodeId, float hz) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? node->vibratoInSfxrUnits(hz) : 0.0;
}
DART_EXPORT float SfxrNode_vibratoInHz(int nodeId, float sfxr) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? node->vibratoInHz(sfxr) : 0.0;
}
DART_EXPORT float SfxrNode_filterFreqInHz(int nodeId, float sfxr) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? node->filterFreqInHz(sfxr) : 0.0;
}

DART_EXPORT float SfxrNode_filterFreqInSfxrUnits(int nodeId, float hz) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    return node ? node->filterFreqInSfxrUnits(hz) : 0.0;
}





DART_EXPORT void SfxrNode_setDefaultBeep(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node)node->setDefaultBeep();
}
DART_EXPORT void SfxrNode_coin(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node)node->coin();
}
DART_EXPORT void SfxrNode_laser(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node)node->laser();
}
DART_EXPORT void SfxrNode_explosion(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node)node->explosion();
}
DART_EXPORT void SfxrNode_powerUp(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node)node->powerUp();
}
DART_EXPORT void SfxrNode_hit(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node)node->hit();
}
DART_EXPORT void SfxrNode_jump(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node)node->jump();
}
DART_EXPORT void SfxrNode_select(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node)node->select();
}



DART_EXPORT void SfxrNode_mutate(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node)node->mutate();
}
DART_EXPORT void SfxrNode_randomize(int nodeId) {
    auto node = std::static_pointer_cast<SfxrNode>(getNode(nodeId));
    if(node)node->randomize();
}

File: lib/src/lab_sound/sfxr_node.dart
import 'audio_context.dart';
import 'audio_node.dart';
import 'audio_param.dart';
import 'lab_sound.dart';

class SfxrNode extends AudioNode {

  late AudioParam attackTime;
  late AudioParam sustainTime;
  late AudioParam sustainPunch;
  late AudioParam decayTime;
  late AudioParam startFrequency;
  late AudioParam minFrequency;
  late AudioParam slide;
  late AudioParam deltaSlide;
  late AudioParam vibratoDepth;
  late AudioParam vibratoSpeed;
  late AudioParam changeAmount;
  late AudioParam changeSpeed;
  late AudioParam squareDuty;
  late AudioParam dutySweep;
  late AudioParam repeatSpeed;
  late AudioParam phaserOffset;
  late AudioParam phaserSweep;
  late AudioParam lpFilterCutoff;
  late AudioParam lpFilterCutoffSweep;
  late AudioParam lpFiterResonance;
  late AudioParam hpFilterCutoff;
  late AudioParam hpFilterCutoffSweep;

  SfxrNode(AudioContext ctx): super(ctx, LabSound().createSfxrNode(ctx.pointer)) {
    attackTime = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_attackTime(this.nodeId));
    sustainTime = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_sustainTime(this.nodeId));
    sustainPunch = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_sustainPunch(this.nodeId));
    decayTime = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_decayTime(this.nodeId));
    startFrequency = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_startFrequency(this.nodeId));
    minFrequency = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_minFrequency(this.nodeId));
    slide = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_slide(this.nodeId));
    deltaSlide = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_deltaSlide(this.nodeId));
    vibratoDepth = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_vibratoDepth(this.nodeId));
    vibratoSpeed = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_vibratoSpeed(this.nodeId));
    changeAmount = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_changeAmount(this.nodeId));
    changeSpeed = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_changeSpeed(this.nodeId));
    squareDuty = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_squareDuty(this.nodeId));
    dutySweep = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_dutySweep(this.nodeId));
    repeatSpeed = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_repeatSpeed(this.nodeId));
    phaserOffset = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_phaserOffset(this.nodeId));
    phaserSweep = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_phaserSweep(this.nodeId));
    lpFilterCutoff = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_lpFilterCutoff(this.nodeId));
    lpFilterCutoffSweep = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_lpFilterCutoffSweep(this.nodeId));
    lpFiterResonance = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_lpFiterResonance(this.nodeId));
    hpFilterCutoff = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_hpFilterCutoff(this.nodeId));
    hpFilterCutoffSweep = AudioParam(this.ctx, this.nodeId, LabSound().SfxrNode_hpFilterCutoffSweep(this.nodeId));
  }

  setStartFrequencyInHz(double value) => LabSound().SfxrNode_setStartFrequencyInHz(nodeId, value);
  setVibratoSpeedInHz(double value) => LabSound().SfxrNode_setVibratoSpeedInHz(nodeId, value);

  double envelopeTimeInSeconds(double sfxrEnvTime) => LabSound().SfxrNode_envelopeTimeInSeconds(nodeId, sfxrEnvTime);
  double envelopeTimeInSfxrUnits(double t) => LabSound().SfxrNode_envelopeTimeInSfxrUnits(nodeId, t);
  double frequencyInSfxrUnits(double hz) => LabSound().SfxrNode_frequencyInSfxrUnits(nodeId, hz);
  double frequencyInHz(double sfxr) => LabSound().SfxrNode_frequencyInHz(nodeId, sfxr);
  double vibratoInSfxrUnits(double hz) => LabSound().SfxrNode_vibratoInSfxrUnits(nodeId, hz);
  double vibratoInHz(double sfxr) => LabSound().SfxrNode_vibratoInHz(nodeId, sfxr);
  double filterFreqInHz(double sfxr) => LabSound().SfxrNode_filterFreqInHz(nodeId, sfxr);
  double filterFreqInSfxrUnits(double hz) => LabSound().SfxrNode_filterFreqInSfxrUnits(nodeId, hz);


  setDefaultBeep() => LabSound().SfxrNode_setDefaultBeep(nodeId);
  coin() => LabSound().SfxrNode_coin(nodeId);
  laser() => LabSound().SfxrNode_laser(nodeId);
  explosion() => LabSound().SfxrNode_explosion(nodeId);
  powerUp() => LabSound().SfxrNode_powerUp(nodeId);
  hit() => LabSound().SfxrNode_hit(nodeId);
  jump() => LabSound().SfxrNode_jump(nodeId);
  select() => LabSound().SfxrNode_select(nodeId);
  mutate() => LabSound().SfxrNode_mutate(nodeId);
  randomize() => LabSound().SfxrNode_randomize(nodeId);

}